query-engine/driver-adapters/executor/src/recordReplay.ts

```diff
@@ -0,0 +1,238 @@
+import {
+  type DriverAdapter,
+  type Query,
+  type Result,
+  type ResultSet,
+  type Transaction,
+  TransactionOptions,
+  ok,
+} from "@prisma/driver-adapter-utils";
+
+export const recordReplay = (adapter: DriverAdapter) => {
+  const recordings = new InMemoryRecordings();
+
+  return {
+    recorder: new RecordDecorator(adapter, recordings),
+    replayer: new ReplayDecorator(adapter, recordings),
+  };
+};
+
+export interface Recordings {
```

Comment:
The design of these abstractions: `Recordings`, `InMemoryRecordings`, `Recorder`, `TxRecorder`,  `Replayer` and `TxReplayer` is subject to change. 

But the point is that there is a parallel hierarchy of Record | Replay, that wrap an adapter adding the record or replay capability. `Recorder` stores computations of queries in `Recordings`, and `Replayer` reads them